src/System.Data.SqlClient/src/System/Data/Common/DbConnectionOptions.cs

```diff
@@ -0,0 +1,709 @@
+            + "\\s*=(?!=)\\s*"                                          // the equal sign divides the key and value parts
+            + "(?<value>"
+            + "(\"([^\"\u0000]|\"\")*\")"                              // double quoted string, " must be quoted as ""
+            + "|"
+            + "('([^'\u0000]|'')*')"                                   // single quoted string, ' must be quoted as ''
+            + "|"
+            + "((?![\"'\\s])"                                          // unquoted value must not start with " or ' or space, would also like = but too late to change
+            + "([^;\\s\\p{Cc}]|\\s+[^;\\s\\p{Cc}])*"                  // control characters must be quoted
+            + "(?<![\"']))"                                            // unquoted value must not stop with " or '
+            + ")(\\s*)(;|[\u0000\\s]*$)"                                // whitespace after value up to semicolon or end-of-line
+            + ")*"                                                      // repeat the key-value pair
+            + "[\\s;]*[\u0000\\s]*"                                     // traling whitespace/semicolons (DataSourceLocator), embedded nulls are allowed only in the end
+        ;
+
+
+        private static readonly Regex ConnectionStringRegex = new Regex(ConnectionStringPattern, RegexOptions.ExplicitCapture | RegexOptions.Compiled);
+#endif
+        private const string ConnectionStringValidKeyPattern = "^(?![;\\s])[^\\p{Cc}]+(?<!\\s)$"; // key not allowed to start with semi-colon or space or contain non-visible characters or end with space
+        private const string ConnectionStringValidValuePattern = "^[^\u0000]*$";                    // value not allowed to contain embedded null
+        private static readonly Regex s_connectionStringValidKeyRegex = new Regex(ConnectionStringValidKeyPattern, RegexOptions.Compiled);
+        private static readonly Regex s_connectionStringValidValueRegex = new Regex(ConnectionStringValidValuePattern, RegexOptions.Compiled);
+
+        // connection string common keywords
+        private static class KEY
+        {
+            internal const string Integrated_Security = "integrated security";
+            internal const string Password = "password";
+            internal const string Persist_Security_Info = "persist security info";
+            internal const string User_ID = "user id";
+        };
+
+        // known connection string common synonyms
+        private static class SYNONYM
+        {
+            internal const string Pwd = "pwd";
+            internal const string UID = "uid";
+        };
+
+        private readonly string _usersConnectionString;
+        private readonly Hashtable _parsetable;
+        internal readonly NameValuePair KeyChain;
+        internal readonly bool HasPasswordKeyword;
+
+        public DbConnectionOptions(string connectionString, Hashtable synonyms)
```

Comment:
In general it'd be nice if the legacy collections could be replaced with the generic ones.